gdpr/models.py

```diff
@@ -30,18 +30,32 @@ def create_from_purpose_slug(self, purpose_slug, source_object, issued_at=None,
             raise KeyError('Purpose with slug {} does not exits'.format(purpose_slug))
 
     def exists_valid_consent(self, purpose_slug, source_object):
-        return LegalReason.objects.filter(
-            source_object_content_type=ContentType.objects.get_for_model(source_object.__class__),
-            source_object_id=str(source_object.pk),
-            purpose_slug=purpose_slug,
-            is_active=True,
-            expires_at__gte=timezone.now(),
-        ).exists()
+        return LegalReason.objects.filter_source_instance_active_non_expired(
+            source_object).filter(purpose_slug=purpose_slug).exists()
+
+
+class LegalReasonQuerySet(models.QuerySet):
+
+    def filter_active(self):
+        return self.filter(is_active=True)
+
+    def filter_non_expired(self):
+        return self.filter(expires_at__gte=timezone.now())
+
+    def filter_active_and_non_expired(self):
+        return self.filter_active().filter_non_expired()
+
+    def filter_source_instance(self, source_object):
+        return self.filter(source_object_content_type=ContentType.objects.get_for_model(source_object.__class__),
```

Comment:
Tohle není potřeba, dá se filtrovat přímo podle objektu. `filter(source_object=source_object)`, takže celá ta funkce podle mě postrádá smysl...

Reply:
ne to není pravda nedá se to. Tohle GenericField neumí

https://docs.djangoproject.com/en/2.1/ref/contrib/contenttypes/#django.contrib.contenttypes.fields.GenericForeignKey.for_concrete_model

Due to the way GenericForeignKey is implemented, you cannot use such fields directly with filters (filter() and exclude(), for example) via the database API

Reply:
OK, máš pravdu, spletl jsem si to s tím že tam můžeš ten objekt přímo uložit.